src/pilot/assign_role.py

```diff
@@ -296,48 +296,116 @@ def get_raid_controller_ids(drac_client):
 
-    raid_10_logical_disk = define_single_raid_10_logical_disk(
-        drac_client, raid_controller_ids[0])
+    # Get the drives controlled by the RAID controller
+    raid_cntlr_physical_disks = {}
+    for disk in all_physical_disks:
+        if disk.controller in raid_controllers:
+            if disk.controller in raid_cntlr_physical_disks:
+                raid_cntlr_physical_disks[disk.controller].append(disk)
+            else:
+                raid_cntlr_physical_disks[disk.controller] = [disk]
+
+    boss_controller = _fetch_boss_controller(drac_client, raid_controllers)
+    if boss_controller:
+        if check_cntlr_physical_disks_len(
+           raid_cntlr_physical_disks[boss_controller]):
+            os_logical_disk = define_storage_operating_system_logical_disk(
+                raid_cntlr_physical_disks[boss_controller],
+                drac_client, boss_controller)
+        else:
+            LOG.critical("The BOSS card has only 1 SSD. "
+                         "2 SSDs are needed to configure a RAID 1")
+            return None
+    else:
+        os_logical_disk = define_single_raid_10_logical_disk(
+            drac_client, raid_controllers[0])
 
-    # None indicates an error occurred.
-    if raid_10_logical_disk is None:
+    if os_logical_disk is None:
         return None
 
-    logical_disks = list()
+    os_physical_disk_names = os_logical_disk['physical_disks'] \
+        if 'physical_disks' in os_logical_disk else None
 
-    # Add the disk to the list only if it is not empty.
-    if raid_10_logical_disk:
-        logical_disks.append(raid_10_logical_disk)
+    # Determine the physical disks that remain for RAID10.
+    # A successful call returns a list, which may be empty; otherwise,
+    # None is returned.
+    logical_disks = [os_logical_disk]
+    remaining_physical_disks = list()
+    for raid_controller in raid_controllers:
+        for disk in raid_cntlr_physical_disks[raid_controller]:
+            if disk.id not in os_physical_disk_names:
+                remaining_physical_disks.append(disk.id)
+
+    if remaining_physical_disks:
+        raid_controller = [cntrl for cntrl in raid_controllers
+                           if not drac_client.is_boss_controller(cntrl)]
+        raid_10_logical_disk = define_logical_disk('MAX',
+                                                   '1+0',
+                                                   raid_controller[0],
+                                                   remaining_physical_disks)
+
+        if raid_10_logical_disk:
+            logical_disks.append(raid_10_logical_disk)
 
     return logical_disks
 
 
-def define_compute_logical_disks(drac_client, raid_controller_ids):
-    if len(raid_controller_ids) > 1:
-        LOG.critical(
-            "Found more than one RAID controller:\n  {}".format(
-                "\n  ".join(raid_controller_ids)))
-        return None
+def define_compute_logical_disks(drac_client, raid_controllers):
```

Comment:
this method looks very similar to define_controller_logical_disks and too much of code is duplicating, need to eliminate duplication. Two of the options:

1) merge the method into one, named e.g. define_controller_or_compute_logical_disks and update where they both are called.
2) introduce private helper method, e.g., _define_controller_or_compute_logical_disks and call it from these 2 methods. This way don't have to change all the places where these 2 are called now.

Reply:
FWIW, I prefer option 2. That option makes it less likely that a programmer will introduce an error by not updating all of the callers. Since this script does not have any unit tests, such errors can very easily go undetected.

Reply:
Added _define_controller_or_compute_logical_disks method which defines the compute and controller logical disk.